### ISSUE TITLE
Support for Razer Chroma Addressable (ARGB) LED Controller

### ISSUE THREAD

**Issue by CalcProgrammer1** [openrazer/openrazer#1357]
### Device Info

* Name: Razer Chroma Addressable LED controller
* Product Number: RZ34-021406
* Model Number: 
* Razer Website: https://www.razer.com/gaming-pc-accessories/razer-chroma-addressable-rgb-controller/RZ34-02140600-R3M1

### Device Information

USB ID: 1532:0F1F (interfaces 0, 1, 2)

### Packet Captures

[ChromaAddressableController_SynapseStartup.zip](https://github.com/openrazer/openrazer/files/5624610/ChromaAddressableController_SynapseStartup.zip)

At around 12 seconds I click the reset button and it rescans the number of attached LEDs.  This appears to use standard 90-byte reports:
[ChromaAddressable_ResetLEDSizeAuto.zip](https://github.com/openrazer/openrazer/files/5625872/ChromaAddressable_ResetLEDSizeAuto.zip)

### HID Descriptor

```
adam@adam-linux-desktop:~/Downloads$ sudo usbhid-dump -d 1532:0f1f
001:005:002:DESCRIPTOR         1606852527.819101
 05 0C 09 01 A1 01 06 00 FF 09 02 15 00 25 01 75
 08 95 10 81 00 C0

001:005:001:DESCRIPTOR         1606852527.822074
 05 0C 09 01 A1 01 06 00 FF 09 02 15 00 25 01 75
 08 96 40 01 B1 01 C0

001:005:000:DESCRIPTOR         1606852527.825013
 05 0C 09 01 A1 01 06 00 FF 09 02 15 00 25 01 75
 08 95 5A B1 01 C0
```

Interface 0 - 90 bytes (standard Razer report, existing control seems to work here, usage page 0x0C and usage 0x01)

Interface 1 - 320 bytes (looks to be direct mode packets that can hold the full 80 LEDs per channel RGB data, also 0x0C 0x01)

Interface 2 - 16 bytes (not sure what this is for, also 0x0C 0x01)

### Synapse Screenshots

Supports auto-detection of LED count and enabled ports, but also supports manual LED sizing.  Looks to cap out at a maximum of 80 LEDs per channel.

![image](https://user-images.githubusercontent.com/1051772/100779482-1562ad00-33ce-11eb-8e57-385ee7564d1d.png)

Individual channel brightness options

![image](https://user-images.githubusercontent.com/1051772/100779554-2b706d80-33ce-11eb-816d-1c8ab4df60f0.png)

I just ordered this controller and would like to get it added to OpenRazer/OpenRGB.  My guess is that it's going to have a command to resize each channel in addition to the modes/direct control messages.

**Comment by CalcProgrammer1** [comment]
I started integrating putting it in razeraccessory, using the same code paths as the Chroma HDK, and it is partially working.  Static effects work fine, but it seems to have a larger 320 byte HID packet for direct control.

Modes Tested:

Direct (Custom):
Off: ✅ 
Static: ✅ 
Breathing (Single): ✅ 
Breathing (Dual):
Breathing (Random):
Spectrum Cycle: ✅ 
Wave (Left): ✅ 
Wave (Right): ✅ 

320 byte packet:

![image](https://user-images.githubusercontent.com/1051772/100804016-bb73de80-33f1-11eb-9dc8-e88a323b2d9c.png)

The highlighted bytes go from 0-5 and repeat, so this must be which channel the data is for.  The data seems to be up to 80x RGB data, for direct control of each LED on the channel.

**Comment by CalcProgrammer1** [comment]
The brightness control seems to be writing to 1A-1F in the LED slot.  The function `razer_chroma_extended_matrix_brightness` should work for brightness.

The "Global Brightness" option doesn't seem to change anything on the protocol side, it just sends the same slider value to all 6 channels.

Brightness works in razeraccessory if I set each channel like so:

```
    case USB_DEVICE_ID_RAZER_CHROMA_HDK:
    case USB_DEVICE_ID_RAZER_CHROMA_BASE:
    case USB_DEVICE_ID_RAZER_NOMMO_PRO:
    case USB_DEVICE_ID_RAZER_NOMMO_CHROMA:
        report = razer_chroma_extended_matrix_brightness(VARSTORE, ZERO_LED, brightness);
        break;

    case USB_DEVICE_ID_RAZER_CHROMA_ADDRESSABLE_RGB_CONTROLLER:
        report = razer_chroma_extended_matrix_brightness(VARSTORE, ARGB_CH_1_LED, brightness);
        mutex_lock(&device->lock);
        razer_send_payload(device->usb_dev, &report);
        mutex_unlock(&device->lock);

        report = razer_chroma_extended_matrix_brightness(VARSTORE, ARGB_CH_2_LED, brightness);
        mutex_lock(&device->lock);
        razer_send_payload(device->usb_dev, &report);
        mutex_unlock(&device->lock);

        report = razer_chroma_extended_matrix_brightness(VARSTORE, ARGB_CH_3_LED, brightness);
        mutex_lock(&device->lock);
        razer_send_payload(device->usb_dev, &report);
        mutex_unlock(&device->lock);

        report = razer_chroma_extended_matrix_brightness(VARSTORE, ARGB_CH_4_LED, brightness);
        mutex_lock(&device->lock);
        razer_send_payload(device->usb_dev, &report);
        mutex_unlock(&device->lock);

        report = razer_chroma_extended_matrix_brightness(VARSTORE, ARGB_CH_5_LED, brightness);
        mutex_lock(&device->lock);
        razer_send_payload(device->usb_dev, &report);
        mutex_unlock(&device->lock);

        report = razer_chroma_extended_matrix_brightness(VARSTORE, ARGB_CH_6_LED, brightness);
        break;
```

It's a bit messy since the function is really only designed to send one report, but this needs to send 6 reports, one for each channel.  It would be possible to expose individual channel brightness control but the way the interface is now only supports global brightness, so that's what I'm going to implement.

**Comment by z3ntu** [comment]
Can you try if changing brightness works with the ZERO_LED? Also I'd be nice if you could share the pcaps you took.

**Comment by CalcProgrammer1** [comment]
I tried using the same code path as the other devices (just stuck the ARGB controller device ID in every case that the Chroma HDK was also in).  This would use the ZERO_LED.  Unfortunately, it didn't work, the brightness stayed full bright on the Chroma ARGB controller while all of my other Razer gear changed brightness appropriately.  With the snippet above, the Chroma ARGB also changed brightness appropriately.  I haven't had time to continue investigating this but I will provide more pcaps when I get back to it.

**Comment by CalcProgrammer1** [comment]
Function to send the 320-byte ARGB packet:

```
struct razer_argb_report {
    unsigned char report_id;
    unsigned char channel_1;
    unsigned char channel_2;
    unsigned char pad;
    unsigned char last_idx;
    unsigned char color_data[315];
};

int razer_send_argb_msg(struct usb_device* usb_dev, unsigned char channel, unsigned char size, void const* data)
{
    uint request = HID_REQ_SET_REPORT; // 0x09
    uint request_type = USB_TYPE_CLASS | USB_RECIP_INTERFACE | USB_DIR_OUT; // 0x21
    uint value = 0x300;
    int len;

    struct razer_argb_report report;

    printk(KERN_WARNING "razer_send_argb_msg: Channel: %d Size: %d.", channel, size);

    if(channel < 5)
    {
        report.report_id = 0x04;
    }
    else
    {
        report.report_id = 0x84;
    }

    report.channel_1 = channel;
    report.channel_2 = channel;

    report.pad = 0;

    report.last_idx = size - 1;

    memcpy(report.color_data, data, size * 3);

    char * buf = kmemdup(&report, sizeof(report), GFP_KERNEL);

    // Send usb control message
    len = usb_control_msg(usb_dev, usb_sndctrlpipe(usb_dev, 0),
        request,            // Request
        request_type,       // RequestType
        value,              // Value
        0x01,               // Index
        buf,                // Data
        sizeof(report),     // Length
        USB_CTRL_SET_TIMEOUT);

    // Wait
    //usleep_range(wait_min, wait_max);

    if (len != sizeof(report))
        printk(KERN_WARNING "razer driver: Device data transfer failed. len = %d", len);

    return ((len < 0) ? len : ((len != size) ? -EIO : 0));
}
```

**Comment by paxos** [comment]
Is there anything I can do to help moving this forward?

**Comment by z3ntu** [comment]
Fixed by #1929
